Streamline route queries

devknoll opened the pull request:
See #104 and #20. This might also conflict a little with #112

Just  wanted to throw this up before I go too far down any rabbit holes. The goal of this PR is to allow route queries to be defined like:
```js
viewer: () => Relay.QL`
  query { viewer }
`,
```
without having to specify `${Component.getFragment('viewer')}` mechanically.

The current change works (for some definition of working) but lacks tests and I'm still wrapping my head around the Relay internals. That being said:
* Does this seem like the correct approach? Any gotchas with variables I'm missing?
* How to we want to handle the two old formats? Should we add some deprecation notices now?
* Any other questions/comments?

.

༼ つ ◕_◕ ༽つ

josephsavona commented on src/container/__tests__/getRelayQueries-test.js:
```diff
@@ -112,7 +112,7 @@ describe('getRelayQueries', () => {
     BadRoute.routeName = 'BadRoute';
     BadRoute.queries = {
-      first: () => Relay.QL`fragment on Node{id}`,
+      first: (Component) => Relay.QL`fragment on Node{id}`,
```
Curious why this test has to change?

josephsavona commented on src/query/buildRQL.js:
```diff
@@ -96,9 +98,21 @@ var buildRQL = {
+          [fragment],
+          query.metadata,
+          query.name
+        );
```
This looks good. FYI the plan is to replace the `GraphQL.*` classes with plain objects and we try to avoid using `new GraphQL.*`, but there's no alternative here.

josephsavona commented:
Be sure to add a `babel-relay-plugin` fixture to ensure that this new syntax will compile without validation errors. We had to add an exception to skip one rule for mutations, and I suspect the same rule may cause issues with the below, because input variables are undefined. See [here](https://github.com/facebook/relay/blob/master/scripts/babel-relay-plugin/src/GraphQLDocumentTransformer.js#L106-L112)

```
// mail fail (current) validation bc no input variables defined
foo: () => Relay.QL`query { node($id) }
```

devknoll commented:
Thanks @josephsavona -- actually, I'm pleased that none of my changes are done in a _total_ vacuum :wink: they all at least get tested on my production app too, since that's where I usually run into an issue first :smile: So I would have caught that! 

Any thoughts with regards to the now two old deprecated formats? Does it still make sense for the signature here to be `() => Relay.QL` (maybe variables require it?) rather than the regular return like mutations?

josephsavona commented:
We have code internally that uses the deprecated variant, and in general functions are preferred for lazy evaluation. The `() => Relay.QL...` pattern feels appropriate here; mutations only use regular functions because of `class` syntax.

josephsavona commented on src/query/__tests__/buildRQL-test.js:
```diff
@@ -188,5 +188,32 @@ describe('buildRQL', () => {
+        }
+      `;
+      var node = buildRQL.Query(
+        builder, MockContainer,
```
separate lines

devknoll commented:
Whoops, thanks @josephsavona. #58 can't come soon enough ;-) Let me know if this is likely to be accepted and I'll add doc updates to the PR

josephsavona commented on src/container/getRelayQueries.js:
```diff
@@ -54,7 +54,8 @@ function getRelayQueries(
         Component,
-        Object.keys(route.params)
+        Object.keys(route.params),
+        Component.getFragment(fragmentName)
```
It's important to pass the route params by default to the fragment, as the user can't manually do so otherwise.

```
Component.getFragment(fragmentName, route.params)
```

josephsavona replied:
We should also add a test case in `getRelayQueries-test` that ensures route params are passed through properly.

josephsavona commented on src/query/buildRQL.js:
```diff
@@ -98,6 +99,17 @@ var buildRQL = {
         node = queryBuilder(Component, variables, RelayQL);
       } else {
```
Let's be more explicit about when the new logic will take effect in order to avoid surprising behavior and have stronger invariants about the expected outcome:
- Only auto-inject the fragment when `queryBuilder` is a function with exactly zero arguments (anything else could be the user doing `(Component) => Relay.QL...`) - so make this a 3rd `else if` clause.
- When the above case is true, check the following conditions:

```javascript
if (GraphQL.isQuery(node) && node.fragments.length === 0) {
  if (!node.fields.every(field => field.isScalar())) {
    invariant(
      false,
      'RelayQL: Expected query `%s` to be empty. For example, use `node(id: $id)`, ' +
      'not `node(id: $id) { ... }`.'
      query.name
    );
  }
  var rootCall = ...
  ...
}
```

This prevent

josephsavona commented:
My apologies for not getting to a more thorough review of this earlier. Please see comments, but this is a good direction. :+1:

devknoll commented:
No worries, thanks for the feedback!

devknoll commented:
> Let's be more explicit about when the new logic will take effect in order to avoid surprising behavior 

Ahhh, had something like that originally. Fixed.

> `field.isScalar()`

`GraphQLFieldNode`s don't have an `isScalar`, so I'm checking for # of child fields directly instead.

> RelayQL: Expected query

I changed this to `buildRQL` since it seems to follow the conventions more as far as I can tell.

> `query.name`

Wasn't sure if you wanted the name from the route/fragment or the root call. Went with the latter, let me know.

> We should also add a test case in getRelayQueries-test that ensures route params are passed through properly.

Let me know if the new test addresses your concerns!

@josephsavona Thanks, and sorry for the delay! I'll get the doc changes in soon too.

josephsavona commented on src/query/__tests__/buildRQL-test.js:
```diff
@@ -188,5 +188,55 @@ describe('buildRQL', () => {
+        builder,
+        MockContainer,
+        ['id'],
+        MockContainer.getQuery()
```
`MockContainer.getFragment('<name>')`

josephsavona commented on src/query/__tests__/buildRQL-test.js:
```diff
@@ -188,5 +188,55 @@ describe('buildRQL', () => {
+          builder,
+          MockContainer,
+          [],
+          MockContainer.getQuery()
```
same as above, `getFragment(...)`

josephsavona commented on src/query/buildRQL.js:
```diff
@@ -81,7 +83,8 @@ var buildRQL = {
     Component: any,
-    variableNames: VariableNames
+    variableNames: VariableNames,
+    fragment: GraphQL.QueryFragment
```
`fragmentReference: RelayFragmentReference`

josephsavona commented on src/query/buildRQL.js:
```diff
@@ -96,8 +99,29 @@ var buildRQL = {
+            rootCall.name,
+            rootCall.value,
+            node.fields,
+            [fragment],
```
`[fragmentReference],`

devknoll commented:
:white_check_mark: `getQuery` -> `getFragment`
:white_check_mark: `fragment` -> `fragmentReference`
:white_check_mark: `GraphQL.QueryFragment` -> `RelayFragmentReference`

I also went ahead and cleaned up the rest of the calls to `getQuery` in buildRQL-test.js, and added a line to the changelog.

cc @josephsavona Thanks!

josephsavona commented:
@devknoll awesome! thanks again for working on this. i'm working on syncing this and your other change.

josephsavona commented on src/container/__tests__/getRelayQueries-test.js:
```diff
@@ -122,4 +122,16 @@ describe('getRelayQueries', () => {
+
+    expect(MockPageContainer.getFragment.mock.calls.length).toBe(4);
+    expect(MockPageContainer.getFragment.mock.calls[0][1]).toBe(params);
+  })
```
semicolon

devknoll replied:
Shoot, sorry!

josephsavona commented on src/container/getRelayQueries.js:
```diff
@@ -54,7 +54,8 @@ function getRelayQueries(
         Component,
-        Object.keys(route.params)
+        Object.keys(route.params),
+        Component.getFragment(fragmentName, route.params)
```
`s/fragmentName/queryName/g` - to match the name change in #181

josephsavona commented on src/query/buildRQL.js:
```diff
@@ -96,8 +100,29 @@ var buildRQL = {
+          if (!node.fields.every(field => field.fields.length === 0)) {
+            invariant(
+              false,
+              'buildRQL: Expected query `%s` to be empty. For example, use ' +
```
Let's use `'Relay.QL: ...` for consistency with other messages.

josephsavona commented on src/query/__tests__/buildRQL-test.js:
```diff
@@ -180,5 +178,55 @@ describe('buildRQL', () => {
+          MockContainer.getFragment('foo')
+        );
+      }).toFailInvariant(
+        'buildRQL: Expected query `viewer` to be empty. For example, use ' +
```
`'Relay.QL:...`

devknoll commented:
Should be good to go now @josephsavona, thanks!

josephsavona commented:
@devknoll this is awesome. thanks!

taion commented:
When using this syntax, the query parameters will not get forwarded to the fragment as variables, right?

devknoll commented:
`route.params` is always passed to `Component.getFragment` with this syntax

taion commented:
Ah, I see. Cool! Is the old syntax to be considered deprecated?

devknoll commented:
I don't think so. Passing different params to `getFragment` seems like a valid thing to want to do, but @josephsavona could comment on longer term plans.

taion commented:
Ah, cool. On closer inspection it also looks like https://github.com/facebook/relay/pull/115/files#diff-b64b5d2eefab1f19d7cf0a21d6546db7R107 gives me pretty good backward compatibility anyway. Very nice.

devknoll commented:
Glad to hear that. Let me know. The more I think about it, the more I'm /facepalming for not just doing this in the babel plugin :smile:

taion commented:
Well, if you change it... let me know.

The one thing I lose out on here is that, when I proxy the fragment for the React Router integration using the current trick, I can restrict the variables passed into the fragment to just the ones that are relevant to the current route (i.e. only the URL parameters for the current route and its ancestors, and the query parameters for the current route): https://github.com/relay-tools/react-router-relay/blob/v0.4.2/src/RouteAggregator.js#L54-L55

With the shorthand, all fragments will receive as variables all parameters on the Relay route. Probably not a big deal in practice. I'm still going to use the shorthand, anyway.
